pkg/controller/association/service_account.go

```diff
@@ -0,0 +1,309 @@
+	esSecret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      elasticsearchSecretName.Name,
+			Namespace: elasticsearchSecretName.Namespace,
+			Labels:    labels,
+		},
+		Data: map[string][]byte{
+			esuser.ServiceAccountTokenNameField: []byte(fullyQualifiedName),
+			esuser.ServiceAccountHashField:      []byte(token.Hash),
+		},
+	}
+	_, err := reconciler.ReconcileSecret(client, esSecret, &es)
+	return err
+}
+
+func ReconcileServiceAccounts(
+	ctx context.Context,
+	client k8s.Client,
+	es esv1.Elasticsearch,
+	commonLabels map[string]string,
+	applicationSecretName types.NamespacedName,
+	elasticsearchSecretName types.NamespacedName,
+	serviceAccount ServiceAccountName,
+	applicationName string,
+	applicationUID types.UID,
+) error {
+	tokenName := tokenName(applicationSecretName.Namespace, applicationName, applicationUID)
+	token, err := reconcileApplicationSecret(ctx, client, es, applicationSecretName, commonLabels, tokenName, serviceAccount)
+	if err != nil {
+		return err
+	}
+	return reconcileElasticsearchSecret(ctx, client, es, elasticsearchSecretName, commonLabels, *token)
+}
+
+// getCurrentApplicationToken returns the current token from the application Secret, or nil if the content of the Secret is not valid.
+func getCurrentApplicationToken(es *esv1.Elasticsearch, secretName string, secretData map[string][]byte) *Token {
+	if len(secretData) == 0 {
+		log.V(1).Info("secret is empty", "es_name", es.Name, "namespace", es.Namespace, "secret", secretName)
+		return nil
+	}
+	result := &Token{}
+	if value := getFieldOrNil(es, secretName, secretData, esuser.ServiceAccountTokenNameField); value != nil && len(*value) > 0 {
+		result.TokenName = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, ServiceAccountTokenValueField); value != nil && len(*value) > 0 {
+		result.Token = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, esuser.ServiceAccountHashField); value != nil && len(*value) > 0 {
+		result.Hash = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, ServiceAccountNameField); value != nil && len(*value) > 0 {
+		result.ServiceAccountName = *value
+	} else {
+		return nil
+	}
+
+	return result
+}
+
+func getFieldOrNil(es *esv1.Elasticsearch, secretName string, secretData map[string][]byte, fieldName string) *string {
+	data, exists := secretData[fieldName]
+	if !exists {
+		log.V(1).Info(fmt.Sprintf("%s field is missing in service account token Secret", fieldName), "es_name", es.Name, "namespace", es.Namespace, "secret", secretName)
+		return nil
+	}
+	fieldValue := string(data)
+	return &fieldValue
+}
+
+var prefix = [...]byte{0x0, 0x1, 0x0, 0x1}
+
+// newApplicationToken generates a new token for a given service account.
+func newApplicationToken(serviceAccountName ServiceAccountName, tokenName string) (*Token, error) {
+	secret := common.RandomBytes(64)
+	hash, err := pbkdf2Key(secret)
```

Comment:
> Initially I had my reservations about reimplementing the `elasticsearch-service-token`CLI here.

This was actually discussed out of band:
> [...] the elasticsearch-service-tokens CLI tool is an easy way to create new file-backed service tokens. But we designed the tokens in a way that it is easy to create them with external tooling as well. This could be useful because you can generate them on the fly without any "feedback" channel [...]

Reply:
Maybe just a sentence in our docs?